### PR TITLE
github issue 01307

### LINKED ISSUE #1307: bigip_provision rejecting valid provisioning options
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

bigip_provision
##### ANSIBLE VERSION

```
ansible 2.7.10
  config file = /home/sxw455/.ansible.cfg
  configured module search path = [u'/home/sxw455/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/sxw455/.local/lib/python2.7/site-packages/ansible
  executable location = /home/sxw455/.local/bin/ansible
  python version = 2.7.15rc1 (default, Nov 12 2018, 14:31:15) [GCC 7.3.0]
```

##### PYTHON VERSION

```
Python 2.7.15rc1
```

##### BIGIP VERSION

```
Sys::Version
Main Package
  Product     BIG-IP
  Version     13.1.1
  Build       0.0.4
  Edition     Final
  Date        Fri Jul 20 17:55:49 PDT 2018
```

##### OS / ENVIRONMENT

```
Ubuntu 18.0.4
```
##### SUMMARY

```
The following provisioning options are rejected in a call:
Level=None (required to de-provision any module; mentioned in the level description, but not available as a value for Level)
Mgmt (would require levels=small, medium, or large)
CGNat (exists, but level values of Enabled or Disabled aren't there)
AAM (not available)
URLDB (not available)
```
##### STEPS TO REPRODUCE

```
See Gist: https://gist.github.com/C0missar/0c4329b8a7562d6f048e5a39b0d622d2
``

<!--- You can also paste gist.github.com links for larger files -->

##### EXPECTED RESULTS
```

That the Mgmt, AAM, CGNat, and URLDB modules would be provisioned
```

##### ACTUAL RESULTS
<!--- What actually happened? If possible run with high verbosity (-vvvv) -->
<!--- Paste verbatim command output between quotes below -->
```
All four provisioning requests were rejected with invalid values.
See Gist: https://gist.github.com/C0missar/0c4329b8a7562d6f048e5a39b0d622d2
```

### PR DESCRIPTION
**Pull request by crosbygw**
fixes #1307 
- adds support for urldb module
- adds support for sslo module
- adds support for cgnat module when using tmos 15.0
- adds support for modifying mgmt reserved memory
- updated unit test
- created new integration tasks
- added setup and teardown tasks to default big-ip module configuration for integration test

### CODE REVIEW

**Comment by crosbygw** [comment]
adding unit and integration test results.
[1307 test results.zip](https://github.com/F5Networks/f5-ansible/files/3383635/1307.test.results.zip)

File: library/modules/bigip_provision.py [hunk 1]
```diff
@@ -51,21 +54,37 @@
...
+      - Use Large for configurations containing more than 2000 objects, or
+        more specifically, for any configuration that exceeds 1000 objects
+        per 2 GB of installed memory. Changing the Management C(mgmt) size
+        after initial provisioning causes a reprovision operation
+    type: str
+    choices:
+      - small
+      - medium
+      - large
+      - integer between 0 and maximum available ram.
```
Comment (wojtek0806) [inline library/modules/bigip_provision.py]: you are missing version added, you needs something like this, but in our case the version is 2.9:

https://github.com/F5Networks/f5-ansible/blob/devel/library/modules/bigip_device_traffic_group.py#L72

File: library/modules/bigip_provision.py [hunk 2]
```diff
@@ -51,21 +54,37 @@
...
     default: nominal
+  memory:
+    description:
+      - Sets additional memory for Management Module.
+      - Options include small, medium, large, or integer between 0 and maximum available ram.
+      - Use Large for configurations containing more than 2000 objects, or
+        more specifically, for any configuration that exceeds 1000 objects
+        per 2 GB of installed memory. Changing the Management C(mgmt) size
+        after initial provisioning causes a reprovision operation
+    type: str
```
Comment (wojtek0806) [inline library/modules/bigip_provision.py]: you are missing version added, you needs something like this, but in our case the version is 2.9:

https://github.com/F5Networks/f5-ansible/blob/devel/library/modules/bigip_device_traffic_group.py#L72

File: library/modules/bigip_provision.py (hunk 1 above)
Comment (wojtek0806) [inline library/modules/bigip_provision.py]: can you explain this :) cause the choices are not part of arg_spec so this needs to go, if you have non standard values like this, just indicate what users should put in in the documentation, also you can add some sanity checks in the ModuleParameters. DOCUMENTATION string always reflects ArgumentSpec class contents.

File: library/modules/bigip_provision.py [hunk 3]
```diff
@@ -234,19 +310,26 @@ def exec_module(self):
             changed = self.present()
         elif state == "absent":
             changed = self.absent()
-
-        changes = self.changes.to_return()
```
Comment (wojtek0806) [inline library/modules/bigip_provision.py]: why this change ?

**Comment by wojtek0806** [comment]
once you have made changes rebase :)

File: library/modules/bigip_provision.py (hunk 3 above)
Comment (crosbygw) [reply library/modules/bigip_provision.py]: Module did not use to have reportable changes but now does. Modified changes variable to now run through ReportableChanges function.
